python/pyspark/ml/tuning.py

```diff
@@ -255,18 +257,24 @@ def _fit(self, dataset):
         randCol = self.uid + "_rand"
         df = dataset.select("*", rand(seed).alias(randCol))
         metrics = [0.0] * numModels
+
+        pool = ThreadPool(processes=min(self.getParallelism(), numModels))
+
         for i in range(nFolds):
             validateLB = i * h
             validateUB = (i + 1) * h
             condition = (df[randCol] >= validateLB) & (df[randCol] < validateUB)
-            validation = df.filter(condition)
+            validation = df.filter(condition).cache()
             train = df.filter(~condition)
-            models = est.fit(train, epm)
-            for j in range(numModels):
-                model = models[j]
+
+            def singleTrain(index):
+                model = est.fit(train, epm[index])
                 # TODO: duplicate evaluator to take extra params from input
-                metric = eva.evaluate(model.transform(validation, epm[j]))
-                metrics[j] += metric/nFolds
+                metric = eva.evaluate(model.transform(validation, epm[index]))
+                metrics[index] += metric/nFolds
+
+            pool.map(singleTrain, range(numModels))
```

Comment:
Can we have a benchmark for this? Can Python GIL a problem here to downgrade performance in the end?

Reply:
Do you mean the line `metrics[index] += metric/nFolds` will downgrade perf because of lock issue ? 
I can change code to avoid this. Thanks!

Reply:
The actual fitting and evaluation methods run here might include CPU bound codes. So I am not sure if multithreading here can well boost the performance.

Reply:
Oh, I think this work well. We already have PRs do similar things #19110 and #16774 .